### ISSUE TITLE
CoreOS on AWS EC2 won't get private ip after reboot/update, if docker.service is stated via cloud-config

### ISSUE THREAD

**Issue by yellowmegaman** [coreos/bugs#1635]
# Issue Report #
CoreOS on AWS EC2 won't get private ip after reboot/update, if docker.service is stated via cloud-config

## Bug ##
Launched coreos cluster with cloud-config on ec2.
Cloud config - http://pastebin.com/raw/mZjzbMwB
docker command start is cornerstone here. If specified, instance won't get privip after reboot. If i specify some container, e.g. redis.service instead of specifying docker.service, everything will be ok.

### CoreOS Version ###
1068.10.0 / 1122.3.0 / beta channel

$ cat /etc/os-release
NAME=CoreOS
ID=coreos
VERSION=1068.10.0
VERSION_ID=1068.10.0
BUILD_ID=2016-08-23-0220
PRETTY_NAME="CoreOS 1068.10.0 (MoreOS)"
ANSI_COLOR="1;32"
HOME_URL="https://coreos.com/"

### Environment ###
AWS EC2 us-east-1d HVM/PV

### Expected Behavior ###
CoreOS should just reboot and get ip normally. E.g.:
After reboot, if i choose "get instance screenshot" i should see:
SSH host keys
eth0: privip ipv6

### Actual Behavior ###
After reboot, if i choose "get instance screenshot" i see:
SSH host keys
eth0: ipv6

### Reproduction Steps ###

  1. Start coreos on EC2 with docker.server command: start in cloud config
  2. Reboot / locksmith update

### Other Information ###
AWS ami's used: 
1) ami-9c086a8b from aws marketplace
2) ami-bf5605a8 ami-7e4f1c69 ami-40570457 from coreos docs

Thanks a lot in advance!

**Comment by crawford** [comment]
> CoreOS on AWS EC2 won't get private ip after reboot/update

Is this only seen via the login issue (the screenshot text you describe), or does it actually affect the networking?

**Comment by yellowmegaman** [comment]
No network at all, no ping/ssh/telnet etcetera.

**Comment by mischief** [comment]
@yellowmegaman can you simply not 'start' docker? anything that needs it should be able to socket-activate it.

**Comment by yellowmegaman** [comment]
Actually, i don't have a need to start-only docker. Just thought that it can be leading to other bugs, since network completely breaks after this configuration. If you consider that minor, or rare-case please close.

**Comment by monder** [comment]
Seems interesting issue, but unfortunately I was unable to reproduce it. Is it only after reboot, and does not happen on the first boot? Maybe there are some units in fleet that are messing up the instance?

**Comment by yellowmegaman** [comment]
Fresh-launched coreos with cloud-config posted above - http://pastebin.com/raw/mZjzbMwB. Nothing else. Cluster of 3-4 machines, IIRC even one machine - same behaviour.

**Comment by crawford** [comment]
I successfully booted and rebooted CoreOS-stable-1122.3.0-hvm (ami-17064d77) with the following config:

```yaml
#cloud-config

coreos:
  update:
    reboot-strategy: etcd-lock
  locksmith:
    window-start: Wed 04:00
    window-length: 1h
  units:
    - name: docker.service
      command: start
```

@yellowmegaman is this enough to trigger the failure in your environment?

**Comment by yellowmegaman** [comment]
@crawford Config above working great, not problems caused by reboot. Tried on single node, 1068.10.0 / ami-9c086a8b. So it must be an option config in my submitted cloud-config.

**Comment by crawford** [comment]
I'm guessing you are scheduling something with fleet which is causing the issue. The rest of your config is fine and is known to work.

**Comment by yellowmegaman** [comment]
Tell you what, i'm just starting with coreos, was using successfully flannel + etcd + consul on debian. And  didn't even begin messing with fleet. Config posted above, with docker.service removed, failed also when i just docker run'd cassandra container. All nodes disconnected, rebooted and no network.

Gona try different ami / coreos version and region and write back all steps.

**Comment by crawford** [comment]
If you can write down step-by-step what you are running, I can try to figure out what is going on. I'm not 100% certain I understand what's happening yet.

As a side note, if you are just getting started with fleet, I would suggest looking at kubernetes instead. We are investing resources in that project instead of fleet at this point and the plan is to retire fleet in the not-to-distant future. I would hate to see you invest time and energy only to see that the project is being killed.

**Comment by yellowmegaman** [comment]
Great many thanks for advice about kubernetes. Time indeed is valuable resource.
I've put everything to terraform script to simplify reproducing.

Hope this log isn't ugly. http://pastebin.com/p4tJP5K3

Some screenshots. First one is coreos1 after reboot. Second and third - coreos2 before and after reboot respectively.

![coreos01](https://cloud.githubusercontent.com/assets/3943191/19905477/ff94c0e0-a087-11e6-841f-c599ee9ad666.png)
![coreos02](https://cloud.githubusercontent.com/assets/3943191/19905479/0214551a-a088-11e6-9a26-999be139d594.png)
![coreos02-after-reboot](https://cloud.githubusercontent.com/assets/3943191/19905484/03e0a01a-a088-11e6-9a5f-dd056dece2bc.png)

**Comment by crawford** [comment]
Weird. I was able to repro it with the following config:

```yaml
#cloud-config

coreos:
  update:
    reboot-strategy: etcd-lock
  locksmith:
    window-start: Wed 04:00
    window-length: 1h
  fleet:
    public-ip: "$public_ipv4"
  flannel:
    etcd_prefix: "/coreos.com/network2"
  units:
    - name: etcd2.service
      command: start
    - name: fleet.service
      command: start
    - name: flanneld.service
      drop-ins:
        - name: 50-network-config.conf
          content: |
            [Service]
            ExecStartPre=/usr/bin/etcdctl set /coreos.com/network2/config '{"Network": "10.0.0.0/8", "SubnetLen": 20, "SubnetMin": "10.10.0.0", "SubnetMax": "10.99.0.0", "Backend": { "Type": "vxlan"}}'
      command: start
    - name: nameservers.network
      content: |
        [Network]
        DNS=8.8.8.8
```

**Comment by yellowmegaman** [comment]
Don't know what's happening anymore. Just booted another cluster and rebooted straight away - same result - total loss of connectivity. Same ami/region

**Comment by crawford** [comment]
I'm trying to narrow down that config to figure out what is going on. It looks like an interaction between fleet and flannel (testing on 1185.3.0).

**Comment by yellowmegaman** [comment]
Removing fleet.service didn't help, soo ...

**Comment by crawford** [comment]
This appears to be the minimal repro case (single node):

```yaml
#cloud-config

coreos:
  units:
    - name: etcd2.service
      command: start
    - name: fleet.service
      command: start
    - name: flanneld.service
      drop-ins:
        - name: 50-network-config.conf
          content: |
            [Service]
            ExecStartPre=/usr/bin/etcdctl set /coreos.com/network/config '{"Network": "10.0.0.0/8", "Backend": { "Type": "vxlan"}}'
      command: start
```

If I don't start flannel or fleet, it reboots fine. If I change the flannel backend to UDP, it reboots fine.

**Comment by yellowmegaman** [comment]
Thanks a lot for rooting this out!! Btw, can you please confirm that vxlan indeed is superior in terms of speed to udp?

**Comment by crawford** [comment]
VXLAN has lower latency than UDP, but their throughput should be similar.

**Comment by yellowmegaman** [comment]
Thanks again for having trouble!

**Comment by yellowmegaman** [comment]
Can you please post working config? Still no luck with this one:

`#cloud-config

ssh_authorized_keys:
  - "ssh-rsa 1"
  - "ssh-rsa 2"
coreos:
  update:
    reboot-strategy: etcd-lock
  locksmith:
    window-start: Wed 04:00
    window-length: 1h
  etcd2:
    discovery: "https://discovery.etcd.io/2f2eac348bfb149a59a393f390c9dea3"
    advertise-client-urls: "http://$private_ipv4:2379"
    initial-advertise-peer-urls: "http://$private_ipv4:2380"
    listen-client-urls: "http://0.0.0.0:2379"
    listen-peer-urls: "http://$private_ipv4:2380"
  flannel:
    etcd_prefix: "/coreos.com/network2"
  units:
    - name: etcd2.service
      command: start
    - name: flanneld.service
      drop-ins:
        - name: 50-network-config.conf
          content: |
            [Service]
            ExecStartPre=/usr/bin/etcdctl set /coreos.com/network2/config '{"Network": "10.0.0.0/8", "SubnetLen": 20, "SubnetMin": "10.10.0.0", "SubnetMax": "10.99.0.0", "Backend": { "Type": "udp", "Port": 7890}}'
      command: start
    - name: nameservers.network
      content: |
        [Network]
        DNS=8.8.8.8`

**Comment by yellowmegaman** [comment]
Got no luck with official docs also.
https://coreos.com/flannel/docs/latest/flannel-config.html

**Comment by yellowmegaman** [comment]
Any news/temp workaround?

**Comment by crawford** [comment]
I did a couple more experiments on this one. After the first boot, if I clear flanneld from the rkt image cache, the second boot works fine (it would fail before). This leads me to believe that this is a timing issue. The fact that the image takes time to download gives this enough time to come up properly. I also tried skipping the cloud-config entirely and just enabling the units (so they run even earlier in the boot). Interestingly, that works just fine. I'd be willing to bet that there is some strange interaction going on between coreos-cloudinit (it will issue `systemctl daemon-reload` with this particular config) and the other components.

My recommendation is to use Ignition on AWS (we are no longer working on coreos-cloudinit). Ignition gives you much more control over the system, with a simpler config, and boots faster (about 40% in my case). This config (from the minimal repro case) works:

```json
{
  "ignition": { "version": "2.0.0" },
  "systemd": {
    "units": [
      {
        "name": "etcd2.service",
        "enable": true
      },
      {
        "name": "fleet.service",
        "enable": true
      },
      {
        "name": "flanneld.service",
        "enable": true,
        "dropins": [{
          "name": "50-network-config.conf",
          "contents": "[Service]\nExecStartPre=/usr/bin/etcdctl set /coreos.com/network/config '{\"Network\": \"10.0.0.0/8\", \"Backend\": { \"Type\": \"vxlan\"}}'"
        }]
      }
    ]
  }
}
```

**Comment by yellowmegaman** [comment]
@crawford Thanks a lot again!

Would be really nice to have some list of actively worked-on coreos components =)
E.g. ignition vs cloud-config, k8s vs fleet.

**Comment by crawford** [comment]
>  Would be really nice to have some list of actively worked-on coreos components

I know... Hang tight, we will be making some announcements shortly.

Since Ignition is the solution to this problem (like so many others), can we close this out?

**Comment by yellowmegaman** [comment]
Think so. But i always thought that cloud-config and stuff like that ( ignition ) is only used to bootstrap system/config, not to be run after reboot/twice etc. So even don't understand how they solve it.

Let's close then, gona try ignition.
And again - thanks for investigating!

**Comment by crawford** [comment]
Canonical's CloudInit and CoreOS's Ignition only run once. Unfortunately, early on when we built coreos-cloudinit, we decided that it would run on every boot (big mistake). Ignition solves the issue because it allows systemd to start everything as specified in the units. Dependencies and ordering are respected. coreos-cloudinit, on the other hand, is effectively it's own init system. It serially executes the config and starts services one by one, so it doesn't respect unit ordering. Additionally, coreos-cloudinit calls `systemctl daemon-reload` and restarts networkd in order to work around the fact that it runs relatively late in the boot process. There are just so many work-arounds that it causes stability problems.

Thanks for your patience and let me know if you have any trouble with Ignition.
